### PR TITLE
Added wharf-core configs

### PR DESCRIPTION
**Pull request by jilleJr**
- Added wharf-core/pkg/config loading
- Changed to use Config throughout instead of env vars
- Added backward compatability with previous environment variables

### CODE REVIEW

**Review by Alexamakans — CHANGES_REQUESTED** [review]
You can completely ignore the addition of the `lookupSeveralEnv` function, it was a spur of the moment thing. Left it in in case you think it looks nice 👍🏻

File: config.go [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+	if cfg.DisableSSL, err = lookupOptionalEnvBool("RABBITMQDISABLESSL", cfg.DisableSSL); err != nil {
+		return err
+	}
+	if cfg.ConnAttempts, err = lookupOptionalEnvUInt64("RABBITMQCONNATTEMPTS", cfg.ConnAttempts); err != nil {
+		return err
+	}
+	return nil
+}
+
+func lookupOptionalEnvBool(name string, fallback bool) (bool, error) {
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply config.go:485]: That got to be the weirdest abuse of Go's pointer mechanics I've seen. Though I'm not strictly against it. The usage became very slick.

Would like to flip around the keys though, so it's something like this instead:

```go
lookupSeveralEnv(map[*string]string{
	&cfg.Username:   "RABBITMQUSER",
	&cfg.Password:   "RABBITMQPASS",
	&cfg.Host:       "RABBITMQHOST",
	&cfg.Port:       "RABBITMQPORT",
	&cfg.VHost:      "RABBITMQVHOST",
	&cfg.QueueName:  "RABBITMQNAME",
})
```

But I'll accept your suggestions and refactor it.

File: config.go [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+	DB   DBConfig
+	MQ   MQConfig
+
+	// InstanceID may be an artibrary string that is used to identify different
+	// Wharf installations from each other. Needed when you use multiple Wharf
+	// installations in the same environment, such as the same Kubernetes
+	// namespace or the same Jenkins instance, to let Wharf know which builds
+	// belongs to which Wharf installation.
+	//
+	// For backward compatability, that may be removed in the next major release
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply config.go]: Nice catch! 1000th time I've misspelled this word.

File: pkg/env/env.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	return nil // can't fail, but just for consitency with the other funcs
+}
+
+// BindMultiple updates the Go variables via the pointers with the values of the
+// environment variables, if set and not empty, for each respective pair in
+// the map.
+//
+// An error is returned if it failed to get any of the environment variables for
+// the mappings.
```
Comment (Alexamakans) [inline pkg/env/env.go:26]: Should maybe change `mappings` to `bindings` to stay consistent with the function names?

Comment (jilleJr) [reply pkg/env/env.go:26]: Good idea!

**Review by Alexamakans — APPROVED** [review]
Nice 👍🏻

**Review by fredx30 — COMMENTED** [review]
I expected a default config file to be committed to the repo with the defaults set. Its much simpler to change a config than to invent one that conforms to a set standard where description is scarce.

**Comment by jilleJr** [comment]
> I expected a default config file to be committed to the repo with the defaults set. Its much simpler to change a config than to invent one that conforms to a set standard where description is scarce.

I thought the reference documentation would be enough, as the types are heavily commented there.

What I like with it being in Go code instead of YAML is that we get compile time safety. I find a YAML file hard to maintain as it easily gets outdated if we're not careful, while the config.go file won't let the tests run if they fail to compile.

A markdown or YAML generator would actually be the best. Something that generates the YAML and environment variable names for each setting, as well as the potential default values. Similar to how we generate our https://github.com/iver-wharf/wharf-helm/blob/master/charts/wharf-helm/README.md file

**Comment by fredx30** [comment]
> > I expected a default config file to be committed to the repo with the defaults set. Its much simpler to change a config than to invent one that conforms to a set standard where description is scarce.
> 
> I thought the reference documentation would be enough, as the types are heavily commented there.
> 
> What I like with it being in Go code instead of YAML is that we get compile time safety. I find a YAML file hard to maintain as it easily gets outdated if we're not careful, while the config.go file won't let the tests run if they fail to compile.
> 
> A markdown or YAML generator would actually be the best. Something that generates the YAML and environment variable names for each setting, as well as the potential default values. Similar to how we generate our https://github.com/iver-wharf/wharf-helm/blob/master/charts/wharf-helm/README.md file

Im still not sure how this gets delivered but i make some assumptions to demonstrate the problem im foreseeing. 
Assuming 
1. the plan to deliver this is through our quay hosted images
2. the quay hosted images contain compiled code in the form of binaries
3. this config file is contained inside the binary file
Then:
Changes to this file seems practically impossible. To change this config you would then have to
1. download the code
2. create your own image build steps for the new settings
3. host to your own container registry
4. write your own helm charts
This seems to void the value of our quay images and helm charts.

I would want a config file to be accessible such that a settings change is as simple as a container file bind top override the config file. Yaml or json config would accmplish this goal much better IMO.

**Comment by jilleJr** [comment]
> Im still not sure how this gets delivered but i make some assumptions to demonstrate the problem im foreseeing.
> Assuming
> 
> 1. the plan to deliver this is through our quay hosted images
> 2. the quay hosted images contain compiled code in the form of binaries
> 3. this config file is contained inside the binary file
> 
> Then:
> Changes to this file seems practically impossible. To change this config you would then have to
> 
> 1. download the code
> 2. create your own image build steps for the new settings
> 3. host to your own container registry
> 4. write your own helm charts
> 
> This seems to void the value of our quay images and helm charts.
> 
> I would want a config file to be accessible such that a settings change is as simple as a container file bind top override the config file. Yaml or json config would accmplish this goal much better IMO.

If it's the config settings for the user that you're worried about then I'm happy to tell you that for a user to add their configs they have some options, as can be seen here:

https://github.com/iver-wharf/wharf-api/blob/70661e4fe6f63dbdacf0378e0b8fb6d4b4deb500/config.go#L350-L358

First, it uses the hardcoded DefaultConfig from the config.go file as a base, and then it merges all the other config file sources on top of that, one of which is `wharf-api-config.yml` in the current working directory, but also if they supply `WHARF_CONFIG` environment variable then that file path is also loaded.

If for say someone hosts Wharf in Kubernetes, then they could mount a [ConfigMap](https://kubernetes.io/docs/concepts/configuration/configmap/) object into the container with the `wharf-api` binary, and then configure Wharf via a secondary resource instead of relying on environment variables.

Changing the default values should not be something the user is expected to do. I agree with you on that part.

Would also like to direct you to these comments:

https://github.com/iver-wharf/wharf-api/blob/70661e4fe6f63dbdacf0378e0b8fb6d4b4deb500/config.go#L11-L30

The link in the README.md file, i.e. this:

https://github.com/iver-wharf/wharf-api/blob/70661e4fe6f63dbdacf0378e0b8fb6d4b4deb500/README.md#L22-L27

Links to that same comment, but rendered over at pkg.go.dev

**Review by fredx30 — APPROVED** [review]
I think i see the larger picture. 👍 
The problem i though i saw does not seem to actually be a problem. I now understand the overrides are still primarily done via env and config files and the code is just defaults. The config generators might be on a future todolist.

**Pull request merged**